File: service_state/vs2008_cpp/TestService/TestStateA.cpp
#include "Common/Locker.h"
#include "Common/Service.h"
#include "Common/TimerManager.h"
#include "TestService.h"
#include "TestStateA.h"

TestStateA::TestStateA()
    : TestState("TestStateA")
{
}

TestStateA::~TestStateA()
{
}

void TestStateA::event1(Service& sv, Event& ev)
{
	printf("TestStateA::event1()\n");
	TimerManager::getInstance().setTimer(*((TestService*)&sv), "test", 1, 1000*2);
	sv.changeState("TestStateB");
}

void TestStateA::event2(Service& sv, Event& ev)
{
	printf("TestStateA::event2()\n");
}

File: service_state/vs2008_csharp/ServiceState/State/TestStateA.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ServiceState.State
{
    class TestStateA : TestState
    {
        public TestStateA() : base("TestStateA")
        {
        }

        public override void EventA(Common.Service service, Common.Event ev)
        {
            System.Threading.Thread.Sleep(1000 * 1);
            ((Service.TestService)service).SendMessage(System.Text.Encoding.ASCII.GetBytes("EventB"));
            service.ChangeState("TestStateB");
        }
    }
}
